python/ray/_private/test_utils.py

```diff
@@ -220,14 +221,28 @@ def run_string_as_driver(driver_script: str, env: Dict = None, encode: str = "ut
         stderr=subprocess.STDOUT,
         env=env,
     )
+    out = ""
+
+    def catch_output(stdout):
+        for line in iter(stdout.readline, b""):
```

Comment:
Is `stdout.readline` threadsafe? Also, is it possible this for loop terminates (reaches EoF) when stdout is written slowly? E.g., there's no more stdout, for loop terminates, and then new output is written to stdout